### PR TITLE
wireguard: allow not storing private keys in world-readable /nix/store

### PR DESCRIPTION
**Pull request by aristidb**
###### Motivation for this change

The wireguard module always used to put the private keys in the Nix Store, which is world-readable.

I changed the module to allow specifying a file instead, which will be read at runtime only.

###### Things done

- [X] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [N/A] macOS
   - [N/A] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by aristidb** [comment]
CC @ericsagnes @zx2c4

File: nixos/modules/services/networking/wireguard.nix [hunk 1]
```diff
@@ -23,8 +23,22 @@ let
 
       privateKey = mkOption {
         example = "yAnz5TF+lXXJte14tji3zlMNq+hd2rYUIgJBgB3fBmk=";
-        type = types.str;
-        description = "Base64 private key generated by wg genkey.";
+        type = with types; nullOr str;
+        default = null;
+        description = ''
+          DEPRECATED BECAUSE IT PUTS A WORLD-READABLE COPY OF THE KEY IN /nix/store!
```
Comment (zx2c4) [inline nixos/modules/services/networking/wireguard.nix]: For some people, this might not be a problem. Rather than depreciating this, I'd keep both supported, but make them mutually exclusive.

File: nixos/modules/services/networking/wireguard.nix [hunk 2]
```diff
@@ -153,35 +182,43 @@ let
...
         description = "WireGuard Tunnel - ${name}";
         after = [ "network.target" ];
         wantedBy = [ "multi-user.target" ];
+        script = builtins.concatStringsSep "\n" (lib.flatten([
+          values.preSetup
+
+          ''
+          ${ipCommand} link del dev "${name}" || true
+          ${ipCommand} link add dev ${name} type wireguard
+          ${wgCommand} setconf ${name} /dev/stdin <<EOF
```
Comment (aristidb) [inline nixos/modules/services/networking/wireguard.nix]: I switched from script to ExecStart because ExecStart did not like my heredoc. It is like this because with a heredoc I could avoid having to create a temporary file and having to make sure it has locked down permissions.

File: nixos/modules/services/networking/wireguard.nix [hunk 3]
```diff
@@ -91,7 +105,21 @@ let
         example = "rVXs/Ni9tu3oDBLS4hOyAUAa1qTWVA3loR8eL20os3I=";
         type = with types; nullOr str;
         description = ''
-          base64 preshared key generated by wg genpsk. Optional,
+          DEPRECATED BECAUSE IT PUTS A WORLD-READABLE COPY OF THE KEY IN /nix/store!
```
Comment (zx2c4) [inline nixos/modules/services/networking/wireguard.nix]: For some people, this might not be a problem. Rather than depreciating this, I'd keep both supported, but make them mutually exclusive.

File: nixos/modules/services/networking/wireguard.nix [hunk 4]
```diff
@@ -134,15 +162,16 @@ let
...
-    PrivateKey = ${values.privateKey}
+    PrivateKey = ${if values.privateKeyFile != null then "$(cat ${values.privateKeyFile})" else values.privateKey}
     ${optionalString (values.listenPort != null)   "ListenPort = ${toString values.listenPort}"}
 
     ${concatStringsSep "\n\n" (map (peer: ''
     [Peer]
     PublicKey = ${peer.publicKey}
-    ${optionalString (peer.presharedKey != null) "PresharedKey = ${peer.presharedKey}"}
+    ${optionalString (peer.presharedKeyFile != null) "PresharedKey = $(cat ${peer.presharedKeyFile})"}
+    ${optionalString (peer.presharedKey != null && peer.presharedKeyFile == null) "PresharedKey = ${peer.presharedKey}"}
```
Comment (zx2c4) [inline nixos/modules/services/networking/wireguard.nix]: This logic here is different and more complex than the logic above for privatekey.

File: nixos/modules/services/networking/wireguard.nix [hunk 5]
```diff
@@ -153,35 +182,43 @@ let
   wgCommand = "${pkgs.wireguard}/bin/wg";
 
   generateUnit = name: values:
+    assert (values.privateKey != null || values.privateKeyFile != null);
```
Comment (zx2c4) [inline nixos/modules/services/networking/wireguard.nix]: I'm glad to see you error out if both are set.

File: nixos/modules/services/networking/wireguard.nix [hunk 6]
```diff
@@ -153,35 +182,43 @@ let
...
     nameValuePair "wireguard-${name}"
       {
         description = "WireGuard Tunnel - ${name}";
         after = [ "network.target" ];
         wantedBy = [ "multi-user.target" ];
+        script = builtins.concatStringsSep "\n" (lib.flatten([
```
Comment (zx2c4) [inline nixos/modules/services/networking/wireguard.nix]: Why did you change this from ExecStart to script? Just curious...

**Review by zx2c4 — COMMENTED** [review]
See inline changes. Needs fixes. Please do not depreciate other fields.

File: nixos/modules/services/networking/wireguard.nix (hunk 2 above)
Comment (Mic92) [inline nixos/modules/services/networking/wireguard.nix]: I would use `wg set` instead of creating a wireguard configuration file. Then a key file can be used directly without the need for here-doc. The old attributes can be still used in this scheme using command substitution `<(echo $key)`

File: nixos/modules/services/networking/wireguard.nix (hunk 5 above)
Comment (aristidb) [inline nixos/modules/services/networking/wireguard.nix]: Actually I don't - this just checks that at least one of the two is set. But you're right, I should.

File: nixos/modules/services/networking/wireguard.nix (hunk 6 above)
Comment (Mic92) [inline nixos/modules/services/networking/wireguard.nix]: ExecStart maps to systemd's `ExecStart`, which is not executed by a shell, but will be exec-ed directly. The author uses shell features here. The `script` argument will create a dedicated shell script, which is placed into an `ExecStart` line.

**Comment by Mic92** [comment]
ping @ericsagnes for later testing of the module.

File: nixos/modules/services/networking/wireguard.nix (hunk 2 above)
Comment (aristidb) [inline nixos/modules/services/networking/wireguard.nix]: @zx2c4 Are there any downsides to using `wg set`?

File: nixos/modules/services/networking/wireguard.nix (hunk 5 above)
Comment (zx2c4) [inline nixos/modules/services/networking/wireguard.nix]: Erm, right.

You should error out if both privateKey and privateKeyFile are set, or if both presharedKey and presharedKeyFile are set.

File: nixos/modules/services/networking/wireguard.nix (hunk 2 above)
Comment (zx2c4) [inline nixos/modules/services/networking/wireguard.nix]: It doesn't make a difference in terms of WireGuard.

It might make a difference in terms of argv stack space though if you have thousands of peers. So if you go the wg set route, I'd recommend you first set up the device:

```
wg set wg0 private-key ... listen-port ...
```

And then set up each peer separately:

```
wg set wg0 peer ABCDEF= endpoint ... preshared-key ...
wg set wg0 peer FAA1234= endpoint ... preshared-key ...
...
```

Also, keep in mind that @Mic92's suggestion of <(echo whatever) requires bash and won't work with generic sh.

Comment (Mic92) [inline nixos/modules/services/networking/wireguard.nix]: The script will be always executed by bash. This should not be an issue.

File: nixos/modules/services/networking/wireguard.nix [hunk 7]
```diff
@@ -26,9 +26,10 @@ let
         type = with types; nullOr str;
         default = null;
         description = ''
-          DEPRECATED BECAUSE IT PUTS A WORLD-READABLE COPY OF THE KEY IN /nix/store!
-
           Base64 private key generated by wg genkey.
+
+          WARNING: Consider using privateKeyFile instead if you don't
```
Comment (zx2c4) [inline nixos/modules/services/networking/wireguard.nix]: I'm not much of a fun of this warning.

If you really insist on having a warning like this, don't use all caps, and say "do not" instead of "don't."

But meh.

File: nixos/modules/services/networking/wireguard.nix [hunk 8]
```diff
@@ -182,7 +186,8 @@ let
   wgCommand = "${pkgs.wireguard}/bin/wg";
 
   generateUnit = name: values:
-    assert (values.privateKey != null || values.privateKeyFile != null);
+    # exactly one way to specify the private key must be set
+    assert (values.privateKey != null) != (values.privateKeyFile != null);
```
Comment (zx2c4) [inline nixos/modules/services/networking/wireguard.nix]: This is good. But you need something similar for preshared keys: only one of the keys may be specified, but not both.

File: nixos/modules/services/networking/wireguard.nix (hunk 2 above)
Comment (Mic92) [inline nixos/modules/services/networking/wireguard.nix]: I did not considered the use case, where one has 1000s of peers. Then using `setconf`, would be probably faster.

File: nixos/modules/services/networking/wireguard.nix [hunk 9]
```diff
@@ -134,14 +164,17 @@ let
...
+    PrivateKey = ${if values.privateKeyFile != null then "$(cat ${values.privateKeyFile})" else values.privateKey}
     ${optionalString (values.listenPort != null)   "ListenPort = ${toString values.listenPort}"}
 
-    ${concatStringsSep "\n\n" (map (peer: ''
+    ${concatStringsSep "\n\n" (map (peer:
+    assert (peer.presharedKeyFile != null) != (peer.presharedKey != null);
+    ''
     [Peer]
     PublicKey = ${peer.publicKey}
+    ${optionalString (peer.presharedKeyFile != null) "PresharedKey = $(cat ${peer.presharedKeyFile})"}
```
Comment (zx2c4) [inline nixos/modules/services/networking/wireguard.nix]: For PrivateKey, you do `PrivateKey = ${if values.privateKeyFile != null then "$(cat ${values.privateKeyFile})" else values.privateKey}`, which is nice since it's clear and one line. However, for PresharedKey, you do two lines.

Can you decide on one strategy so that this is consistent?

File: nixos/modules/services/networking/wireguard.nix (hunk 6 above)
Comment (zx2c4) [inline nixos/modules/services/networking/wireguard.nix]: Where is the script created and what is its file mode? If it's world readable, then this PR doesn't actually fix anything.

File: nixos/modules/services/networking/wireguard.nix (hunk 9 above)
Comment (zx2c4) [inline nixos/modules/services/networking/wireguard.nix]: also, since you're in bash, you can do `$(< path/to/file)` to avoid execing cat.

File: nixos/modules/services/networking/wireguard.nix [hunk 10]
```diff
@@ -134,14 +164,17 @@ let
 
   };
 
-  generateConf = name: values: pkgs.writeText "wireguard-${name}.conf" ''
+  generateConf = name: values: ''
     [Interface]
-    PrivateKey = ${values.privateKey}
+    PrivateKey = ${if values.privateKeyFile != null then "$(cat ${values.privateKeyFile})" else values.privateKey}
```
Comment (zx2c4) [inline nixos/modules/services/networking/wireguard.nix]: (Since you're in bash, you can do `$(< path/to/file)` to avoid execing cat.)

File: nixos/modules/services/networking/wireguard.nix [hunk 11]
```diff
@@ -23,8 +23,23 @@ let
...
+          Base64 private key generated by wg genkey.
+
+          Warning: Consider using privateKeyFile instead if you do not
+          want to store the key in the world-readable Nix store.
+        '';
+      };
+
+      privateKeyFile = mkOption {
+        example = "/private/wireguard_key";
+        type = with types; nullOr str;
```
Comment (ericsagnes) [inline nixos/modules/services/networking/wireguard.nix]: `nullOr path` feels more appropriate.

File: nixos/modules/services/networking/wireguard.nix [hunk 12]
```diff
@@ -91,7 +106,22 @@ let
...
+
+          Warning: Consider using presharedKeyFile instead if you do not
+          want to store the key in the world-readable Nix store.
+        '';
+      };
+
+      presharedKeyFile = mkOption {
+        default = null;
+        example = "/private/wireguard_psk";
+        type = with types; nullOr str;
```
Comment (ericsagnes) [inline nixos/modules/services/networking/wireguard.nix]: `nullOr path` feels more appropriate.

File: nixos/modules/services/networking/wireguard.nix (hunk 9 above)
Comment (ericsagnes) [inline nixos/modules/services/networking/wireguard.nix]: Note that there is a nix function to get a file contents, `lib.strings.fileContents`.

**Review by ericsagnes — COMMENTED** [review]
Added a few nitpicks comments.
I am quite busy and don't use wireguard at the moment so I might not be able to test the module any soon, sorry for the trouble.

File: nixos/modules/services/networking/wireguard.nix (hunk 9 above)
Comment (zx2c4) [inline nixos/modules/services/networking/wireguard.nix]: Yes but the whole purpose is to **avoid** putting it inside any world-readable files, so actually using `fileContents` would be a complete disaster.
